pkg/featuremanager/feature_manager.go

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package featuremanager
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"log/slog"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
+	"github.com/cilium/statedb"
+
+	"github.com/cilium/cilium/pkg/dynamicconfig"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+type FeatureStatus struct {
+	Feature Feature
+	Enabled bool
+}
+
+type FeatureManager interface {
+	Append(feature Feature, hook cell.HookInterface)
+	List() []Feature
+
+	start(feature Feature, ctx context.Context) error
+	stop(feature Feature, ctx context.Context) error
+	process(status []FeatureStatus, ctx context.Context) error
+}
+
+type DefaultFeatureManager struct {
+	featureLifecycle FeatureLifecycle
+	logger           *slog.Logger
+}
+
+func (d *DefaultFeatureManager) List() []Feature {
+	return d.featureLifecycle.List()
+}
+
+// Append adds the specified hook to the list of hooks managed for the given feature
+// The lifecycle will be managed by the FeatureManager
+func (d *DefaultFeatureManager) Append(feature Feature, hook cell.HookInterface) {
+	if err := d.featureLifecycle.Append(feature, hook); err != nil {
+		d.logger.Error("Failed to append feature", "feature", feature, logfields.Error, err)
+		return
+	}
+}
+
+func (d *DefaultFeatureManager) start(feature Feature, ctx context.Context) error {
+	if err := d.featureLifecycle.Start(feature, ctx, d.logger); err != nil {
+		d.logger.Error("Failed to start feature", "feature", feature, logfields.Error, err)
+		return err
+	}
+	return nil
+}
+
+func (d *DefaultFeatureManager) stop(feature Feature, ctx context.Context) error {
+	if err := d.featureLifecycle.Stop(feature, ctx, d.logger); err != nil {
+		d.logger.Error("Failed to stop feature", "feature", feature, logfields.Error, err)
+		return err
+	}
+	return nil
+}
+
+func (d *DefaultFeatureManager) process(status []FeatureStatus, ctx context.Context) error {
+	// The implementation leaks some safety features when enabling compute intensive features
+	// see https://github.com/cilium/cilium/issues/34930
+	d.logger.Info("Processing feature status", "features", status)
+
+	var errs error
+	for _, feature := range status {
+		var err error
+		if feature.Enabled {
+			err = d.start(feature.Feature, ctx)
```

Comment:
Shouldn't we use a context with a timeout here? Otherwise a misbehaving feature would prevent all others from being started. Though we might also need to account for `start` not caring about `ctx`.